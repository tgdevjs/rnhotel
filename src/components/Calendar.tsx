import React, { Component } from "react";
import { StyleSheet, View } from "react-native";
import { CalendarList } from "react-native-calendars";
import moment from "moment";

import { DateRange } from "./date-range";
import { ReservationDateRangeType } from "../types";

type Props = {
  endDay: string;
  onSetDates: (dates: ReservationDateRangeType) => void;
  startDay: string;
};

type State = {
  currentDate: string;
  selectedDates: ReservationDateRangeType;
  markedDates: MarkedDates;
};

type MarkedDates = { [key: string]: MarkedCalendarDate };

type MarkedCalendarDate = {
  startingDay: boolean;
  endingDay: boolean;
  disableTouchEvent: boolean;
  color: string;
  textColor: string;
};

type SelectedCalendarDate = {
  dateString: string;
};

const styles = StyleSheet.create({
  calendar: {
    borderTopWidth: 1,
    paddingTop: 5,
    borderBottomWidth: 1,
    borderColor: "#eee",
    height: 350
  },
  text: {
    textAlign: "center",
    borderColor: "#bbb",
    padding: 10,
    backgroundColor: "#eee"
  },
  container: {
    flex: 1
  }
});

export class Calendar extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const { endDay, startDay } = this.props;

    this.state = {
      currentDate: moment().format(moment.HTML5_FMT.DATE),
      selectedDates: { endDay, startDay },
      markedDates: this.getMarkedDates(startDay, endDay)
    };

    this.onDayPress = this.onDayPress.bind(this);
  }

  selectedColor = "#00bfff";
  textColor = "white";

  static getRangeOfDates(start: string, end: string): [string?] {
    const startMoment = moment(start);
    const endMoment = moment(end);
    const now = startMoment.clone();
    const dates: [string?] = [];

    while (now.isSameOrBefore(endMoment)) {
      dates.push(now.format(moment.HTML5_FMT.DATE));
      now.add(1, "days");
    }
    return dates;
  }

  getMarkedDates(start: string, end: string): MarkedDates {
    // Return an object with keys of dates between start and end date
    const { selectedColor, textColor } = this;
    const dates = Calendar.getRangeOfDates(start, end);
    const markedDates: MarkedDates = dates.reduce(
      (previousValue, currentValue) => ({
        ...previousValue,
        [currentValue]: { color: selectedColor, textColor }
      }),
      {}
    );
    markedDates[start].startingDay = true;
    markedDates[end].endingDay = true;

    return markedDates;
  }

  onDayPress({ dateString }: SelectedCalendarDate): void {
    const { selectedColor, textColor } = this;
    const {
      selectedDates: { endDay, startDay }
    } = this.state;

    let newStartDay = dateString;
    let newEndDay = null;

    // Set MarkedDates to selected day
    let markedDates: MarkedDates = {
      [newStartDay]: {
        startingDay: true,
        endingDay: true,
        disableTouchEvent: true,
        color: selectedColor,
        textColor
      }
    };

    if (!endDay && moment(dateString).isAfter(moment(startDay))) {
      // Set Marked Dates to range between startDay and selectedDay
      newStartDay = startDay;
      newEndDay = dateString;
      markedDates = this.getMarkedDates(newStartDay, newEndDay);
    }

    const selectedDates: ReservationDateRangeType = {
      startDay: newStartDay,
      endDay: newEndDay
    };

    this.setState({ selectedDates, markedDates }, () => {
      this.props.onSetDates(selectedDates);
    });
  }

  render() {
    const {
      currentDate,
      selectedDates: { startDay, endDay },
      markedDates
    } = this.state;

    return (
      <View style={styles.container}>
        <DateRange startDay={startDay} endDay={endDay} />
        <CalendarList
          style={styles.calendar}
          current={currentDate}
          onDayPress={this.onDayPress}
          pastScrollRange={0}
          futureScrollRange={50}
          scrollEnabled
          showScrollIndicator
          markedDates={markedDates}
          markingType="period"
          minDate={currentDate}
        />
      </View>
    );
  }
}
