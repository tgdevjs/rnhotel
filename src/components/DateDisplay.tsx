import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import PropTypes from 'prop-types';
import moment from 'moment';

const styles = StyleSheet.create({
  datesContainer: {
    alignItems: 'center',
  },
  dates: {
    flexDirection: 'row',
  },
  dateContainerLeft: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  dateContainerRight: {
    flex: 1,
    flexDirection: 'row',
  },
  dayContainer: {
    justifyContent: 'center',
  },
  dateText: {
    fontSize: 40,
    fontWeight: '600',
  },
});
export class DateDisplay extends React.PureComponent {
  render() {
    const { dateTextStyle, endDay, startDay, style } = this.props;
    const startMoment = moment(startDay);
    const endMoment = moment(endDay);

    return (
      <View style={[styles.datesContainer, style]}>
        <View style={styles.dates}>
          <View style={styles.dateContainerLeft}>
            <Text style={[styles.dateText, dateTextStyle]}>{startMoment.format('DD')}</Text>
            <View style={styles.dayContainer}>
              <Text>{startMoment.format('ddd')}</Text>
              <Text>{startMoment.format('MMM')}</Text>
            </View>
          </View>
          <Text style={[styles.dateText, dateTextStyle]}> - </Text>
          {!endDay ? (
            <View style={styles.dateContainerRight} />
          ) : (
            <View style={styles.dateContainerRight}>
              <Text style={[styles.dateText, dateTextStyle]}>
                {endDay ? endMoment.format('DD') : '  '}
              </Text>
              <View style={styles.dayContainer}>
                <Text>{endMoment.format('ddd')}</Text>
                <Text>{endMoment.format('MMM')}</Text>
              </View>
            </View>
          )}
        </View>
      </View>
    );
  }
}

DateDisplay.propTypes = {
  dateTextStyle: PropTypes.shape({}),
  endDay: PropTypes.string,
  startDay: PropTypes.string.isRequired,
  style: PropTypes.shape({}),
};

DateDisplay.defaultProps = {
  dateTextStyle: null,
  endDay: null,
  style: null,
};
