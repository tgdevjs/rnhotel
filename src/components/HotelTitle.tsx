import React from "react";
import { StyleSheet, Text } from "react-native";

import { colors, fonts } from "../styles";

type Props = {
  style?: object;
  title: string;
};

const styles = StyleSheet.create({
  title: {
    ...fonts.H3,
    color: colors.white,
    textAlign: "center"
  }
});

export const HotelTitle = ({ style, title }: Props) => {
  return <Text style={[styles.title, style]}>{title}</Text>;
};
