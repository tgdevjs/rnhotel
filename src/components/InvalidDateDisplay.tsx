import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import PropTypes from 'prop-types';

const styles = StyleSheet.create({
  Container: {
    alignItems: 'center',
    borderColor: 'lightgray',
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
  },
  text: {
    color: 'dodgerblue',
    fontSize: 20,
    fontWeight: '600',
  },
  messageContainer: {
    flexDirection: 'row',
  },
});
export class InvalidDateDisplay extends React.PureComponent {
  render() {
    const { endDay, startDay } = this.props;

    return (
      <View style={styles.Container}>
        <Text style={styles.text}>Invalid dates</Text>
        <View style={styles.messageContainer}>
          <Text>{startDay}</Text>
          <Text>to</Text>
          <Text>{endDay}</Text>
        </View>
      </View>
    );
  }
}

InvalidDateDisplay.propTypes = {
  endDay: PropTypes.string,
  startDay: PropTypes.string,
};

InvalidDateDisplay.defaultProps = {
  endDay: null,
  startDay: null,
};
