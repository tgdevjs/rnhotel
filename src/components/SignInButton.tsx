import React from "react";
import { StyleSheet, Text, TouchableOpacity } from "react-native";
import { borderRadius, colors, spacing } from "../styles";

type Props = {
  onPress: () => void;
};

const styles = StyleSheet.create({
  signInButton: {
    padding: spacing,
    borderRadius: borderRadius.small,
    backgroundColor: colors.dogerBlue
  },
  signInButtonText: {
    color: colors.white,
    fontWeight: "bold",
    textAlign: "center"
  }
});

export const SignInButton = ({ onPress }: Props) => {
  return (
    <TouchableOpacity style={styles.signInButton} onPress={onPress}>
      <Text style={styles.signInButtonText}>Sign In</Text>
    </TouchableOpacity>
  );
};
