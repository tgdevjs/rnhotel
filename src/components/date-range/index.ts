export * from "./DateRange";
export * from "./Nights";
export * from "./SingleDate";
export * from "./DateRangeError";
