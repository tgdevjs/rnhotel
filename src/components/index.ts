export * from "./Calendar";
export * from "./date-range";
export * from "./HotelTitle";
export * from "./SignInButton";
