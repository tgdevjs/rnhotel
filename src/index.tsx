/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 */

import React from 'react';
import { ApolloProvider } from 'react-apollo';

import { client, restoreCache } from './apollo';
import Navigation from './Navigation';

export default class App extends React.Component {
  state = { isSetupComplete: false };

  async componentDidMount() {
    await restoreCache();
    this.setState({ isSetupComplete: true });
  }

  render() {
    const { isSetupComplete } = this.state;

    if (!isSetupComplete) return null;
    return (
      <ApolloProvider client={client}>
        <Navigation />
      </ApolloProvider>
    );
  }
}
