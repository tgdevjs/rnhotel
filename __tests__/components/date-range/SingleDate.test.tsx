import React from "react";
import { SingleDate } from "../../../src/components";
import moment from "moment";

describe("SingleDate", () => {
  afterEach(() => {
    mockdate.reset();
  });
  it("renders default", () => {
    expect(shallow(<SingleDate />)).toMatchSnapshot();
  });
});
