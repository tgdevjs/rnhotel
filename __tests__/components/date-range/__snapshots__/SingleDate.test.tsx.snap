// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleDate renders default 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 48,
          "fontWeight": "bold",
        },
        undefined,
      ]
    }
  >
    12
  </Text>
  <View
    style={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <Text>
      Sun
    </Text>
    <Text>
      May
    </Text>
  </View>
</View>
`;
